### ISSUE TITLE
Bathymetry being altered in assign_initial_conditions

### ISSUE THREAD

**Issue by jhill1** [thetisproject/thetis#150]
Hi,

Odd problem that seems to occur on one of my simulations. I read in a netcdf file to create the bathymetry/topography. When I look at the output the bathymetry looks ok, except it seems to be clipped at 27m depth, despite the topography going to 40m+ (i.e. -40m). It otherwise looks OK (not skewed of "off" from my mesh).

So I then printed out the minimum in the bathymetry.data.dat (i.e. the highest land) variable at various stages of the simulation and get the following output:

```
Reading bathymetry/topography data from inputs/shetlands_palaeo_bathy_100m.nc
Interpolating bathymetry/topography data onto simulation mesh
Bathymetry/topography min:-196.91967167572423
Bathymetry/topography max:129.63601321824117
-196.91967167572423
before init
-196.91967167572423
before init cond
-196.91967167572423
dt = 4.0
Using time integrator: CrankNicolson
after init cond
26.610915552352306
before start
26.610915552352306
    0     0 T=      0.00 eta norm:     0.1133 u norm:     0.1133  0.00
volume2d rel. error  0.0000e+00
Exporting: uv_2d elev_2d 
Exporting: 
time:  4.0
26.610915552352306
```

The lines of code are:
```python
print("before init")
print(min(bathymetry2d.dat.data))
update_forcings(0.0)
# set initial condition for elevation, piecewise linear function
print("before init cond")
print(min(bathymetry2d.dat.data))
solverObj.assign_initial_conditions(uv=Constant((1.0e-6,0.0)), elev=Constant(1.0e-6))
print("after init cond")
print(min(bathymetry2d.dat.data))
#uv, elev = solverObj.timestepper.solution.split()
# Run model
print("before start")
print(min(bathymetry2d.dat.data))
```

You can see the bathymetry gets altered after assign_initial_conditions. This has been driving me mad for a couple of days now. I've dug around the code and can't see anything obvious. Any ideas from someone who knows the code better?

**Comment by tkarna** [comment]
Thanks for submitting this. Would it be possible for you to share the entire script you're running? The mesh and bathymetry data set would also be useful.

**Comment by jhill1** [comment]
Bathy, mesh and script are here: https://drive.google.com/drive/folders/1kSFF4zD-q8ZhBmlAnxy83qnK-B490bVc?usp=sharing
(Note I've only shared this with @tkarna as the bathy data are propitiatory so can't be shared publicly).

**Comment by tkarna** [comment]
Thanks! This appears to be a non-trivial bug. Somehow related to evaluating UFL expressions. For the time begin, a quick fix is to declare the boundary values with `Constant` instead of a float:
```python
solverObj.bnd_functions['shallow_water'] = {
    physID: {'elev': Constant(0.0)},
    1000: {'un': Constant(0.0)},
}
```

**Comment by jhill1** [comment]
Thanks for the workaround @tkarna :+1:

**Comment by tkarna** [comment]
A fix is in place in branch `fix-2d-elev-bnd`. Could you try that out?

**Comment by tkarna** [comment]
fixed in #151.
